vllm/spec_decode/spec_decode_worker.py

```diff
@@ -478,6 +484,19 @@ def _run_no_spec(self, execute_model_req: ExecuteModelRequest,
                 self.previous_hidden_states.update(
                     execute_model_req.seq_group_metadata_list, hidden_states)
 
+        if not skip_proposer:
+            # For prefill step in proposer, we run the model for N-1 tokens
+            # because Nth token will be processed in the first decode step. For
+            # N-1 tokens, the input should be 0:N-1 hidden states which should
+            # be concatanated with 1:N token (since output of scorer has to be
+            # the input for proposer). Therefore, we shift the hidden states to
+            # align n-1th hidden state with nth token.
+            if sampler_output.prefill_hidden_states is not None:
+                execute_model_req.previous_hidden_states = sampler_output\
+                    .prefill_hidden_states.roll(shifts=1, dims=0)
```

Comment:
Thanks for adding a great comment.

Can this logic be pushed into the draft worker?

Reply:
Ah got it. Nice.

Consider adding a comment explaining why this is here and moving this to a separate helper function (concern here is too much edge-case logic in the main flow makes it hard to read).